crates/rust-analyzer/src/config.rs

```diff
@@ -2199,193 +2302,143 @@ pub enum TargetDirectory {
-                    $({
-                        let field = stringify!($field3);
-                        let ty = stringify!($ty3);
-
-                        (field, ty, &[$($doc3),*], serde_json::to_string(&$default3).unwrap().as_str())
-                    },)*
-                ])
-            }
-
-            #[cfg(test)]
-            fn manual() -> String {
-                manual(&[
-                    $({
-                        let field = stringify!($field3);
-                        let ty = stringify!($ty3);
-
-                        (field, ty, &[$($doc3),*], $default3)
+                        (field, ty, &[$($doc),*], default)
                     },)*
                 ])
             }
         }
 
-        #[test]
-        fn fields_are_sorted() {
-            [$(stringify!($field)),*].windows(2).for_each(|w| assert!(w[0] <= w[1], "{} <= {} does not hold", w[0], w[1]));
-            [$(stringify!($field2)),*].windows(2).for_each(|w| assert!(w[0] <= w[1], "{} <= {} does not hold", w[0], w[1]));
-            // [$(stringify!($field3)),*].windows(2).for_each(|w| assert!(w[0] <= w[1], "{} <= {} does not hold", w[0], w[1]));
+        mod $modname {
+            #[test]
+            fn fields_are_sorted() {
+                let field_names: &'static [&'static str] = &[$(stringify!($field)),*];
+                field_names.windows(2).for_each(|w| assert!(w[0] <= w[1], "{} <= {} does not hold", w[0], w[1]));
+            }
         }
     };
 }
 use _config_data as config_data;
+use _default_str as default_str;
+use _default_val as default_val;
+
+#[derive(Debug, Clone, Serialize)]
+struct ConfigData {
+    #[serde(flatten)]
+    global: GlobalConfigData,
+    #[serde(flatten)]
+    local: LocalConfigData,
+    #[serde(flatten)]
+    client: ClientConfigData,
+}
+
+impl ConfigData {
+    fn from_json(
+        mut json: serde_json::Value,
+        error_sink: &mut Vec<(String, serde_json::Error)>,
+    ) -> ConfigData {
+        ConfigData {
+            global: GlobalConfigData::from_json(&mut json, error_sink),
+            local: LocalConfigData::from_json(&mut json, error_sink),
+            client: ClientConfigData::from_json(&mut json, error_sink),
+        }
+    }
+
+    fn from_toml(
+        mut toml: toml::Table,
+        error_sink: &mut Vec<(String, toml::de::Error)>,
+    ) -> ConfigData {
+        ConfigData {
+            global: GlobalConfigData::from_toml(&mut toml, error_sink),
+            local: LocalConfigData::from_toml(&mut toml, error_sink),
+            client: ClientConfigData::from_toml(&mut toml, error_sink),
+        }
+    }
+
+    fn schema_fields() -> Vec<SchemaField> {
+        let mut fields = Vec::new();
+        GlobalConfigData::schema_fields(&mut fields);
+        LocalConfigData::schema_fields(&mut fields);
+        ClientConfigData::schema_fields(&mut fields);
+        // HACK: sort the fields, so the diffs on the generated docs/schema are smaller
+        fields.sort_by_key(|&(x, ..)| x);
```

Comment:
This is a bit of a hack. We may want three config sections, each being alphabetically sorted, and with a heading and a preamble. Or, alternatively, each documentation entry has an asciidoc tag saying which config class it is in, `global`/`local`/`client`.

Reply:
I thought about having the latter otherwise it is an extra step for users to figure out if the wanted config is a global/local/client.